framework/src/araui/views/ui_view.c

```diff
@@ -0,0 +1,338 @@
+ *
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <string.h>
+#include <pthread.h>
+#include <araui/ui_widget.h>
+#include <araui/ui_view.h>
+#include <araui/ui_commons.h>
+#include "ui_log.h"
+#include "ui_core_internal.h"
+#include "ui_view_internal.h"
+#include "ui_request_callback.h"
+#include "ui_commons_internal.h"
+
+#define UI_VIEW_SET_STATE(v, s)                                 \
+	do {                                                        \
+		if (v) {                                                \
+			(v)->state = s;                                     \
+			if (s == UI_VIEW_STATE_SHOWN) {                     \
+				g_current_view = v;                             \
+			}                                                   \
+			if ((v)->state_changed_cb[s]) {                     \
+				(v)->state_changed_cb[s]((ui_view_t)v);         \
+			}                                                   \
+		}                                                       \
+	} while (false)
+
+static vec_void_t g_view_list;
+static vec_void_t g_view_update_list;
+static ui_rect_t g_rect_mempool[CONFIG_UI_UPDATE_MEMPOOL_SIZE];
+static int g_rect_mempool_idx = 0;
+static ui_view_body_t *g_current_view = UI_NULL;
+
+static void _ui_view_create_func(void *userdata);
+static void _ui_view_destroy_func(void *userdata);
+static ui_rect_t *_ui_view_get_mempool_rect(void);
+
+ui_error_t ui_view_list_init(void)
+{
+	vec_init(&g_view_list);
+
+	return UI_OK;
+}
+
+ui_error_t ui_view_list_deinit(void)
+{
+	ui_view_body_t *view;
+	int iter;
+
+	vec_foreach(&g_view_list, view, iter) {
+		_ui_view_destroy_func((void *)view);
+	}
+	vec_deinit(&g_view_list);
+
+	return UI_OK;
+}
+
+ui_error_t ui_view_update_list_init(void)
+{
+	vec_init(&g_view_update_list);
+
+	return UI_OK;
+}
+
+ui_error_t ui_view_update_list_deinit(void)
+{
+	vec_deinit(&g_view_update_list);
+
+	return UI_OK;
+}
+
+ui_view_t ui_view_create(created_callback created_cb, destroyed_callback destroyed_cb, shown_callback shown_cb, hidden_callback hidden_cb)
+{
+	ui_view_body_t *body;
+
+	if (!ui_is_running()) {
+		UI_LOGE("error: UI framework is not running!\n");
+		return UI_NULL;
+	}
+
+	if (!created_cb) {
+		UI_LOGE("error: invalid parameter!\n");
+		return UI_NULL;
+	}
+
+	body = (ui_view_body_t *)UI_ALLOC(sizeof(ui_view_body_t));
+	if (!body) {
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_view_body_t));
+
+	body->root = (ui_widget_body_t *)ui_widget_create(CONFIG_UI_DISPLAY_WIDTH, CONFIG_UI_DISPLAY_HEIGHT);
+	body->focus = body->root;
+	body->state_changed_cb[UI_VIEW_STATE_CREATED] = created_cb;
+	body->state_changed_cb[UI_VIEW_STATE_SHOWN] = shown_cb;
+	body->state_changed_cb[UI_VIEW_STATE_HIDDEN] = hidden_cb;
+	body->state_changed_cb[UI_VIEW_STATE_DESTROYED] = destroyed_cb;
+
+	if (ui_request_callback(_ui_view_create_func, body) != UI_OK) {
+		UI_FREE(body);
```

Comment:
Don't you need to free body->root here?

Reply:
Please, check all similar situations through the whole files.

Reply:
Fixed